### ISSUE TITLE
Misleading documentation example: `optimal_thread_count = multiprocessing.cpu_count()`

### ISSUE THREAD

**Issue by AXGKl** [ReactiveX/RxPY#521]
https://rxpy.readthedocs.io/en/latest/get_started.html#concurrency

this example is misleading:

```python
# calculate number of CPUs, then create a ThreadPoolScheduler with that number of threads
optimal_thread_count = multiprocessing.cpu_count()
pool_scheduler = ThreadPoolScheduler(optimal_thread_count)
```

Its in since long, but I think it is confusing for python people where thread pool sizes are not related to available CPUs.

**Comment by MainRo** [comment]
Can you elaborate ? The comment of the example clearly states that the code creates a ThreadPool where the number of threads equals the number of cores (cpu).

**Comment by AXGKl** [comment]
Hi, sorry for long waiting. 

`optimal_thread_count = multiprocessing.cpu_count()`

 
And that is in deed how you would do it e.g. in Java to get max throughput at cpu bound stuff. Because there threads spread over CPUs. 

Not so in Python, even with a 1000 threads you can load only one CPU.
Means: CPUs and optimal Threads within *one* python process are not related.

If you want to use more cores, since you are CPU bound, you have to go multiprocessing. By tuning threads you will reach nothing.
 
On the other hand if you are I/O bound or have to sleep, then threads, cooperative or not are the answer.
In his example he sleeps and thats why parallelizing the function calls makes sense. Would he *really* do an intense calculation instead of sleeping than the threads would even slow done the prozess further, due to the context switching overhead, only multiprozessing would be the answer then - his program is NOT multiprocess, it's one process. 

Saying *optimal* thread count for an app is related to available CPUs is **imho** showing the world, that you do not really understand the difference between python threads and system threads  - and we should really delete that from the getting started page or change the wording.

If you need more infos maybe check [here](https://medium.com/towards-artificial-intelligence/the-why-when-and-how-of-using-python-multi-threading-and-multi-processing-afd1b8a8ecca#:~:text=The%20Python%20GIL%20means%20that,using%20multiprocessing%20can%20improve%20performance) (first hit in google about the subject).

**Comment by MainRo** [comment]
Thanks the precisions. The process/multi-process naming is effectively misleading, the example is about multi-threading. We must update this.
However concurrency on multiple python threads is still possible when escaping from the GIL. As explained on the beginning of the example some libraries implemented in native can run concurrently.

**Comment by AXGKl** [comment]
Hi Romain,

> As explained 
yes saw it, valid - same scientific or gaming libs really doctored the GIL out, which is surprisingly easy - but it's not that easy to get then required data structure locking requirements correct - and fast. Good talk minutes of Hettinger about that here: https://youtu.be/9zinZmE3Ogk?t=746.

Means: I guess multiprocessing is still the way to go mainstream wise when you want to load cores for "intense calculations" - and that would prbly be the simplest way to keep the example w/o major rephrasing effort: Spawn processes instead of sleeping and done, then the wording can be mainly kept.

-----

Maybe we should go a step further and explain people how to do

1. in process concurrency for blocking operations
2. multiprocess concurrency for CPU intense operations

within Rx, so that they have a blueprint?

And maybe without separate streams but in-stream (giving up msgs order then), e.g. like this for 1:

```python
my_stream.pipe(
    group_by(req_blocking),
    flat_map(lambda s: s if not s.key else s.pipe(
                                        delay(0, my_async_scheduler),
                                        map(my_blocking_function)))
         
```
and 2. as described in #528 

If you agree and think its worth it I can also write it down and send you a PR, just drop me a note here. 


PS: Excellent blog you have there!

**Comment by MainRo** [comment]
Sure we accept improvements on the doc !
Describing both cases would be great, simply because each one is adapted to different situations. I think that the difficulty here is to find simple enougth examples (and maybe a multiprocess scheduler in rxpy would help...).

**Comment by pablocastilla** [comment]
Hi

Is there some future work to support multiprocessing? I think it could be a very good feature, imagine a process that accepts events from multiple sources and wants to work with them independently in parallel using the whole CPU.

Thanks

**Comment by dbrattli** [comment]
I'm closing this since I don't agree that the docs are misleading. Multiple Python threads can utilize multiple cores when escaping the GIL as when using e.g Numpy and other external libraries.

**Comment by pablocastilla** [comment]
but numpy uses C. We can not do that in rxpy

**Comment by dbrattli** [comment]
Yes, but we can also use numpy using the ThreadPoolScheduler, or use the ProcessPoolExecutor:

```python
import time
from concurrent.futures import ProcessPoolExecutor

import reactivex
from reactivex import operators as ops
from reactivex.observable.observable import Observable

seconds = [5, 1, 2, 4, 3]


def work(timeout: float) -> float:
    time.sleep(timeout)
    return timeout


def output(result: float) -> None:
    print("%d seconds" % result)


with ProcessPoolExecutor(5) as executor:

    def mapper(s: float) -> Observable[float]:
        return reactivex.from_future(executor.submit(work, s))

    reactivex.from_(seconds).pipe(
        ops.flat_map(mapper),
    ).subscribe(output)
```

**Comment by ziodave** [comment]
I am trying to understand more about this, I added the thread name to the above calls

```
def work(timeout: float) -> float:
    print("[%s] waiting %d seconds" % (current_thread().name, timeout))
    time.sleep(timeout)
    return timeout


def output(result: float) -> None:
    print("[%s] %d seconds" % (current_thread().name, result))
```

This is the output:

```
[MainThread] waiting 5 seconds
[MainThread] waiting 1 seconds
[MainThread] waiting 2 seconds
[MainThread] waiting 4 seconds
[MainThread] waiting 3 seconds
[Thread-1] 1 seconds
[Thread-1] 2 seconds
[Thread-1] 3 seconds
[Thread-1] 4 seconds
[Thread-1] 5 seconds
```

Is it correct that the the thread name is always the same? I would expected different names.

**Comment by ziodave** [comment]
Nevermind, I added the `getpid`, I can see the different process id now:

```python
def work(timeout: float) -> float:
    print("[%d:%s] waiting %d seconds" % (getpid(), current_thread().name, timeout))
    time.sleep(timeout)
    return timeout


def output(result: float) -> None:
    print("[%d:%s] %d seconds" % (getpid(), current_thread().name, result))
```

```
[16104:MainThread] waiting 5 seconds
[16105:MainThread] waiting 1 seconds
[16107:MainThread] waiting 2 seconds
[16106:MainThread] waiting 4 seconds
[16109:MainThread] waiting 3 seconds
[16102:Thread-1] 1 seconds
[16102:Thread-1] 2 seconds
[16102:Thread-1] 3 seconds
[16102:Thread-1] 4 seconds
[16102:Thread-1] 5 seconds
```
